### ISSUE TITLE
ExpressionBodySyntax should allow for exclusion on line wrapping

### ISSUE THREAD

**Issue by marianosimone** [arturbosch/detekt#950]
## Expected Behavior
According to the official [Kotlin styleguide](https://android.github.io/kotlin-guides/style.html#expression-functions):

> Expression functions should not wrap to two lines. If an expression function grows to require wrapping, use a normal function body, a return declaration, and normal expression wrapping rules instead.

## Current Behavior

`ExpressionBodySyntax` doesn't take this into account, and reports violations even if there are line wraps:

```
fun doALot(foo: String): String {
    return foo
         .replace("a", "0")
         .replace("b", "1")
}
````

Slightly more complicated is the fact that, sometimes, the expression does fit in a line, but couldn't be made a expression body because it would break another rule (like max line length):

I couldn't put this in a line if my max line length is 50, for example

```
fun doAVeryLittle(foo: String) : String {
    return foo.replace("a", "0")
}
```

**Comment by schalkms** [comment]
You are right. This makes sense to adapt. Thanks for submitting and documenting this!

**Comment by schalkms** [comment]
Closed with e35d184

**Comment by scott-kennedy-ck** [comment]
It's still popping up for me when I have a when statement as the only return expression:

```kotlin
fun foo(bar: Int): Int {
    return when (bar) {
        0 -> 1
        1 -> 2
        else -> 3
    }
}

**Comment by schalkms** [comment]
@scott-kennedy-ck would you mind creating a new issue? This rule needs some more testing, I think.

**Comment by marianosimone** [comment]
I created the follow up issue.. and will be sending a PR shortly

**Comment by marianosimone** [comment]
I created a PR to fix @scott-kennedy-ck 's report.. but I'm still not sure this rule is doing exactly the right thing (so, I wouldn't consider this issue closed).

Back to my original example:

Suppose that my `MaxLineLength` is `50`, and the following function:

    fun doAVeryLittle(foo: String) = foo.replace("a", "0")

As it doesn't fit in one line, I'd have to do:

```
fun doAVeryLittle(foo: String) =
    foo.replace("a", "0")
```

However, the way I interpret Kotlin's styleguide:

> Expression functions should not wrap to two lines. If an expression function grows to require wrapping, use a normal function body, a return declaration, and normal expression wrapping rules instead.

that's not the right thing to do, and I should be doing this instead (which is reported by `ExpressionBodySyntax`:

```
fun doAVeryLittle(foo: String) : String {
    return foo.replace("a", "0")
}
```

Is there any way to make one rule check another's config to accomplish this?

**Comment by arturbosch** [comment]
Well normaly rules do not interact with each other but we could use a hack here:
The interface `ConfigAware` which every rule implements has the `config` property.
This property is the subconfig of the ruleset. As `ExpressionBodySyntax` and `MaxLineLength` are both from the `Style-RuleSet` we could use `config.subConfig("MaxLineLength").valueOr(Default|Null)`.
This is kinda hacky because we need to hardcode the `MaxLineLength-String` or use the `MaxLineLength-Class.simpleName` etc. But it would work.

OR: we provide an additional property to `ExpressionsBodySyntax` which is basically a repetition of the `maxLineLength` property. Negativ: errorprone to change for the user. Positive: he can control this rule better and even say it should be 100 and not 120 like the normal max line length.

What do you think?

**Comment by marianosimone** [comment]
- As a `detekt` user: I prefer the hacky way. There's only 1 configuration I have to take care of, and it perfectly follows the official style guide (it says "one line")
- As a `detekt` collaborator perspective: provide an additional property to configure this rule
- As a "let's make `detekt` as usable for everyone as possible": let's have both:
    - add the new config option
    - if it's not present, default to the one in `MaxLineLength`

**Comment by schalkms** [comment]
I think this rule should not be overloaded with complexity. I also would not want to have weird dependencies between rules. In this case I am okay with this. I prefer @arturbosch 's first option because it's simpler IMHO.
